doc/compiler/optimizer/OptimizerDesignFeatures.md

```diff
@@ -0,0 +1,303 @@
+or alternately due to a method getting sampled frequently during the run.
+
+There is a [JIT sampling thread](https://github.com/eclipse-openj9/openj9/blob/20c2e83228780955fd62e1620160a81e7cbec41e/runtime/compiler/control/CompilationRuntime.hpp#L1359)
+that periodically tracks the method at the top of the execution stack on each application thread and keeps
+sampling statistics for each method.
+
+Depending on the runtime in question, the specific thresholds for both invocation counts and sampling counts
+can be configured. Initial compilations are typically driven by a combination of counting and sampling,
+whereas recompilations typically depend on sampling since the overhead of counting can be significant in
+compiled code.
+
+The OpenJ9 JIT compiler supports both [asynchronous](https://github.com/eclipse-openj9/openj9/blob/20c2e83228780955fd62e1620160a81e7cbec41e/runtime/compiler/control/CompilationRuntime.hpp#L764)
+and synchronous models for performing compilations.
+
+In an asynchronous model, the thread that is requesting compilation of a method continues executing the method
+after queueing it for compilation, whereas in a synchronous model, the requesting thread waits until the
+compilation is done.
+
+Most compilations are typically done asynchronously since it is better to allow work to be done on the
+requesting thread, but synchronous compilations are necessary in some cases when it is not feasible to
+execute the method in any other way (more on this later in the section on speculative optimization).
+
+The compiler also supports doing compilations on [multiple threads](https://github.com/eclipse-openj9/openj9/blob/20c2e83228780955fd62e1620160a81e7cbec41e/runtime/compiler/control/CompilationRuntime.hpp#L974)
+concurrently to generate compiled code more quickly.
+
+Performance is not only about peak throughput and other metrics such as start-up time and the time to ramp
+up to peak throughput can be improved significantly by asynchronous compilation and using multiple threads
+to do compilations.
+
+A mixed mode execution model provides the flexibility to choose the basis for doing compilation.
+
+Sometimes, it becomes necessary to transition mid-invocation from interpreter to compiled code (e.g. when
+the method is only invoked once) and from compiled code to interpreter (e.g. when the JIT compiler does not
+implement some rare edge case or if we choose to skip compiling some paths for performance reasons).
+The OpenJ9 JIT compiler has both capabilities.
+
+The mid-invocation transition from interpreter to compiled code is called [Dynamic Loop Transfer (DLT)](https://github.com/eclipse-openj9/openj9/blob/master/doc/compiler/control/CompilationControl.md#dynamic-loop-transfer)
+and the mid-invocation transition from compiled code to interpreter is called On Stack Replacement (OSR).
```

Comment:
Should this include a reference to https://github.com/eclipse-openj9/openj9/blob/24b14b4743f82a648712bdca71a5b870828e65f9/doc/compiler/hcr/NextGenHCR.md?

Reply:
Good point! I'll include a link pointing to the OSR section: https://github.com/eclipse-openj9/openj9/blob/24b14b4743f82a648712bdca71a5b870828e65f9/doc/compiler/hcr/NextGenHCR.md#osr-implementation-details